File: Dynamic Programming/Edit_Distance.py
import sys,math,random
from heapq import heappush,heappop
from bisect import bisect_right,bisect_left
from collections import Counter,deque,defaultdict
from itertools import permutations

# functions #
MOD = 998244353
MOD = 10**9 + 7
RANDOM = random.randrange(2**62)
def gcd(a,b):
    if a%b==0:
        return b
    else:
        return gcd(b,a%b)
def lcm(a,b):
    return a//gcd(a,b)*b
def w(x):
    return x ^ RANDOM
##

#String hashing : sh, fenwick sortedlist : fsortl, Number : numtheory, SparseTable : SparseTable
#bucket sorted list : bsortl, segment tree(lazy propogation) : SegmentTree, bootstrap : bootstrap
#binary indexed tree : BIT, segment tree(point updates) : SegmentPoint, Convex Hull : hull
#Combinatorics : pnc, Diophantine Equations : dpheq
#Template : https://github.com/OmAmar106/Template-for-Competetive-Programming

def solve():
    #L1 = list(map(int, sys.stdin.readline().split()))
    s = sys.stdin.readline().strip()
    t = sys.stdin.readline().strip()

    m = len(s)
    n = len(t)

    dp = [[0]*(n+1) for i in range(m+1)]
    

    for i in range(m + 1):
        for j in range(n + 1):
            if i == 0:
                dp[i][j] = j
            elif j == 0:
                dp[i][j] = i
            elif s[i - 1] == t[j - 1]:
                dp[i][j] = dp[i - 1][j - 1]
            else:
                dp[i][j] = 1 + min(dp[i][j - 1], dp[i - 1][j], dp[i - 1][j - 1])
    print(dp[m][n])
solve()

File: Dynamic Programming/Edit_Distance.cpp
// Choose cpp when the time
// limit will for sure give tle in pypy
#include <bits/stdc++.h>
using namespace std;
#define int long long
#define INF LLONG_MAX
#define fastio ios::sync_with_stdio(false); cin.tie(0);
#define print(dp) for (auto it : dp){cout<<it<<" ";}cout<<endl;
#define len(dp) dp.size()
#define printf(x) cout << x << endl;
#define printm(map) cout<<"{";for(auto it: map){cout<<it.first<<":"<<it.second<<",";};cout<<"}"<<endl;

// const int MOD = 998244353;
const int MOD = 1e9 + 7;

int gcd(int a, int b) {
    if (b == 0) return a;
    return gcd(b, a % b);
}

int lcm(int a, int b) {
    return a / gcd(a, b) * b;
}

// String hashing: sh/shclass, Number: numtheory, SparseTable: SparseTable
// Segment Tree(lazy propogation): SegmentTree
// binary indexed tree: BIT, Segment Tree(point updates): SegmentPoint, Convex Hull: hull, Trie/Treap: Tries
// Combinatorics: pnc, Diophantine Equations: dpheq, Graphs: graphs, DSU: DSU, Geometry: Geometry, FFT: fft
// Persistent Segment Tree: perseg, FreqGraphs: bgraph
// Template : https://github.com/OmAmar106/Template-for-Competetive-Programming

void solve() {
    string s;
    string s1;
    cin>>s;
    cin>>s1;

    int dp[len(s)][len(s1)];
    if(s[0]==s1[0]){
        dp[0][0] = 0;
    }
    else{
        dp[0][0] = 1;
    }
    for(int i=0;i<len(s);i++){
        for(int j=0;j<len(s1);j++){
            if((i-1>=0) || (j-1>=0)){
                dp[i][j] = INF;
                if(s[i]==s1[j] && (i>=1) &&(j>=1)){
                    dp[i][j] = min(dp[i][j],dp[i-1][j-1]);
                }
                else if((i>=1) && (j>=1)){
                    dp[i][j] = min(dp[i][j],dp[i-1][j-1]+1);
                }
                else if(s[i]==s1[j]){
                    dp[i][j] = min(dp[i][j],max(i,j));
                }
                if(i>=1){
                    dp[i][j] = min(dp[i][j],dp[i-1][j]+1);
                }
                if(j>=1){
                    dp[i][j] = min(dp[i][j],dp[i][j-1]+1);
                }
            }
        }
    }
    printf(dp[len(s)-1][len(s1)-1]);
}

int32_t main() {
    fastio
    int t=1;
    while (t--) solve();
    return 0;
}